ports/litex/supervisor/internal_flash.c

```diff
@@ -0,0 +1,350 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2013, 2014 Damien P. George
+ * Copyright (c) 2019 Lucian Copeland for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#include "supervisor/internal_flash.h"
+
+#include <stdint.h>
+#include <string.h>
+#include <stdbool.h>
+
+#include "extmod/vfs.h"
+#include "extmod/vfs_fat.h"
+#include "py/mphal.h"
+#include "py/obj.h"
+#include "py/runtime.h"
+#include "lib/oofatfs/ff.h"
+
+#include "supervisor/usb.h"
+
+#include "csr.h"
+#include "irq.h"
+
+enum pin {
+    PIN_MOSI = 0,
+    PIN_CLK = 1,
+    PIN_CS = 2,
+    PIN_MISO_EN = 3,
+    PIN_MISO = 4, // Value is ignored
+};
+
+#define NO_CACHE        0xffffffff
+
+static uint8_t  _flash_cache[FLASH_PAGE_SIZE] __attribute__((aligned(4)));
+static uint32_t _flash_page_addr = NO_CACHE;
+static bool     _flash_cache_dirty;
+// -------------------------------------------------------------------------
+// When performing SPI operations, the flash cannot be accessed.  Since we
+// normally execute directly from SPI, this can cause problems.
+// To work around this, we execute from RAM.  This is accomplished by marking
+// functions as being in the section ".ramtext".
```

Comment:
I've been playing tricks with the linker script to accomplish the same.

Should I call RAM "ITCM", place all `.itcm.` sections into SRAM, and modify `linker.h` to add another exception for litex?  I'm not sure if it will be fast enough, but I can try.

Reply:
Sure! That should be fine.

Reply:
This has now been done.